### ISSUE TITLE
Kafka offramp stops sending data to Kafka if the send request times out

### ISSUE THREAD

**Issue by anuprout** [wayfair-tremor/tremor-runtime#306]
**Problem**

It was observed that the Kafka offramp stops sending records to Kafka when the Kafka brokers are responding slow leading to a request time out. It looks like the producer buffer queue  gets filled out quick if Kafka brokers respond slow. As a result the producer completely stops sending any new records . 

**Steps**

1. Start a Kafka offramp with default value for  **queue.buffering.max.messages**.
2. Keep sending records to the offramp with large volume
3. Inject a netwrok delay to Kafka brokers using any network emulator such as [netem](http://manpages.ubuntu.com/manpages/trusty/man8/tc-netem.8.html) 

**Possible Solution(s)**

Currently Tremor is using **rdkafka::producer::FutureProducer** to send record to Kafka with 1 ms  wait on the producer buffer queue to be freed. It can be changed to either no-wait or use **send_result** instead of **send** function . Use of send_result is preferred as it will return error immediately in case of failure in enqueuing. 

```rust
impl Offramp for Kafka {
    // TODO
    fn on_event(&mut self, codec: &Box<dyn Codec>, _input: String, event: Event) -> Result<()> {
        for value in event.value_iter() {
            let raw = codec.encode(value)?;
            let mut record = FutureRecord::to(&self.topic);
            record = record.payload(&raw);
            //TODO: Key
            if let Some(ref k) = self.key {
                task::spawn(self.producer.send(record.key(k.as_str()), 1));
            } else {
                task::spawn(self.producer.send(record, 1));
            };
        }
        Ok(())
    }
```
**Notes**

Output of `rustup --version`:
Output of `rustup show`:
Output of `tremor-server --version`:0.8.0

(edit/hg: added code block for readability)

**Comment by Licenser** [comment]
Hi Anup! Good catch :)!

It's perfect timing that you found this, we're currently working on circuit breaker and guaranteed delivery and that will require to revisit the Kafka offramp! Knowing about this issue ahead of getting there wills safe some time :D.

I'll link this to the current PR in the area and look at it as part of the offramp overhaul.

If this issue is time critical let us know and we'll look into options to back port that changes from dev to 0.8.*

**Comment by anuprout** [comment]
Thanks @Licenser . This is critical for us now as we see this issue quite frequently. Please see if the change can be added as a hot fix to 0.8.*.

**Comment by Licenser** [comment]
Heya :D thanks for the quick feedback! If it's critical then something will make it in 0.8 hotfixes one way or the other :), I'll put that on the plate first thing next week to look into and see how quick we can resolve fixed.
